File: 10_UniView/Python_Tk/UniData.py
# Module UniData.py
# Reads and stores Unicode UCD file UnicodeData.txt (currently use version from Unicode 13)
# Provides official name and category for each valid codepoint. Not restricted to BMP.
# Can't name it UnicodeData since there is already a module with this name in Python
# Use Python 3.6/3.7 features
#
# 2018-09-06    PV
# 2020-09-03    PV      1.1: Unicode 13, Tcl support >0xFFFF, list contains UTF-8 and UTF-16
# 2020-12-31    PV      1.2: Added Scripts.txt, GetUnknown(cp)

from types import MappingProxyType
from typing import Dict, List
from dataclasses import dataclass


MAXCODEPOINT = 0x10FFFF

# # Python 3.6, an immutable namedtuple (import collections)
# Can't use immutable anymore, since scripts are read in a 2nd pass after records are created
# Good news, @dataClass from Python 3.7 is mutable!
# class CharacterRecord(NamedTuple):
#     Codepoint: int
#     Name: str
#     Category: str
#     Script: str

def GetCodepointHexa(Codepoint: int) -> str:
    return f'U+{Codepoint:04X}'

def GetUTF8String(Codepoint: int) -> str:
    if Codepoint <= 0x7F:
        return f"{Codepoint:02X}"
    if Codepoint <= 0x7FF:
        return f"{0xC0 + Codepoint // 0x40:02X} {0x80 + Codepoint % 0x40:02X}"
    if Codepoint <= 0xFFFF:
        return f"{0xE0 + (Codepoint // 0x40) // 0x40:02X} {0x80 + (Codepoint // 0x40) % 0x40:02X} {0x80 + Codepoint % 0x40:02X}"
    if Codepoint <= 0x1FFFFF:
        return f"{0xF0 + ((Codepoint // 0x40) // 0x40) // 0x40:02X} {0x80 + ((Codepoint // 0x40) // 0x40) % 0x40:02X} {0x80 + (Codepoint // 0x40) % 0x40:02X} {0x80 + Codepoint % 0x40:02X}"
    return "?{Codepoint}?"

def GetUTF16String(Codepoint: int) -> str:
    return f'{Codepoint:04X}' if Codepoint <= 0xD7FF or 0xE000 <= Codepoint <= 0xFFFF else f'{0xD800 + ((Codepoint - 0x10000) >> 10):04X} {0xDC00 + (Codepoint & 0x3ff):04X}'


@dataclass
class CharacterRecord:
    Codepoint: int
    Name: str
    Category: str
    Script: str

    @property
    def CodepointHexa(self) -> str:
        return GetCodepointHexa(self.Codepoint)

    @property
    def UTF16(self) -> str:
        return GetUTF16String(self.Codepoint)

    @property
    def UTF8(self) -> str:
        return GetUTF8String(self.Codepoint)


# Read information file UnicodeVersion.txt
UnicodeVersion = ''
with open('UnicodeVersion.txt', encoding='utf-8') as uv:
    UnicodeVersion = uv.readline()

_codepoint_map: Dict[int, CharacterRecord] = {}

# Read name and category from UnicodeData.txt
with open('UnicodeData.txt', encoding='utf-8') as ud:
    for line in (ud):
        fields: List[str] = line.split(';')
        codepoint = int(fields[0], 16)
        char_name = fields[1]
        char_category = fields[2]

        # Special name overrides
        if codepoint == 28:
            char_name = "CONTROL - FILE SEPARATOR"
        else:
            if codepoint == 29:
                char_name = "CONTROL - GROUP SEPARATOR"
            else:
                if codepoint == 30:
                    char_name = "CONTROL - RECORD SEPARATOR"
                else:
                    if codepoint == 31:
                        char_name = "CONTROL - UNIT SEPARATOR"
                    else:
                        if codepoint < 32 or codepoint >= 0x7f and codepoint < 0xA0:
                            char_name = "CONTROL - " + (fields[10] if len(fields[10]) > 0 else fields[0][2:])

        is_range = char_name.endswith(', First>')
        if is_range:    # add all characters within a specified range
            char_name = char_name.replace(', First>', '').replace('<', '').upper()  # remove range indicator from name
            line = next(ud)
            fields = line.split(';')
            end_char_code = int(fields[0], 16)
            if not fields[1].endswith(', Last>'):
                raise Exception('Expected end-of-range indicator.')
            for code_in_range in range(codepoint, end_char_code+1):
                _codepoint_map[code_in_range] = CharacterRecord(
                    code_in_range, f'{char_name} - {code_in_range:X}', char_category, 'Unknown')
        else:
            _codepoint_map[codepoint] = CharacterRecord(codepoint, char_name, char_category, 'Unknown')

# Read script from Scripts.txt
with open('Scripts.txt', encoding='utf-8') as ud:
    for line in (ud):
        p = line.find('#')
        if p >= 0:
            line = line[:p]
        if len(line) == 0 or line == '\n':
            continue
        fields = line.split(';')
        codepoint_range = fields[0].strip()
        script_name = fields[1].strip()
        p = codepoint_range.find('..')
        if p < 0:
            codepoint = int(codepoint_range, 16)
            _codepoint_map[codepoint].Script = script_name
        else:
            from_codepoint = int(codepoint_range[:p], 16)
            to_codepoint = int(codepoint_range[p+2:], 16)
            for codepoint in range(from_codepoint, to_codepoint+1):
                _codepoint_map[codepoint].Script = script_name


# For public access, readonly dict of CharacterRecord
CharacterRecords: MappingProxyType[int, CharacterRecord] = MappingProxyType(_codepoint_map)


# Use another internal dict for efficient mapping name -> codepoint, though it's not used much in this app
_name_map: Dict[str, int] = {cr.Name.lower(): cr.Codepoint for cr in _codepoint_map.values()}


def GetCodepointFromName(name: str) -> int:
    cp = _name_map.get(name.lower(), -1)
    return cp


def AsString(cp: int) -> str:
    if cp<0:
        print('cp:', cp)
        pass
    return chr(cp)


# Value returned for unsupported codepoints
unknown = CharacterRecord(-1, 'Unknown character', '??', 'Unknown')

# def GetUnknown(cp: int) -> CharacterRecord:
#     return CharacterRecord(cp, f"Unassigned codepoint - {cp:04X}", '??', "Unknown")


def GetName(cp: int) -> str:
    cr = _codepoint_map.get(cp, None)
    return cr.Name if cr else f"Unassigned codepoint - {cp:04X}"

def GetCategory(cp: int) -> str:
    cr = _codepoint_map.get(cp, None)
    return cr.Category if cr else "??"

def GetScript(cp: int) -> str:
    cr = _codepoint_map.get(cp, None)
    return cr.Script if cr else "Unknown"

def IsValidCodepoint(cp: int) -> bool:
    return cp in _codepoint_map

def IsSurrogate(cp: int) -> bool:
    return 0xD800 <= cp <= 0xDFFF



# Tests
if __name__ == '__main__':
    # Commented out, not needed once it works.  And it raises an error in mypy (attempt to modify a readony variable)
    # # Check that CharacterRecord is immutable, not the case with @dataclass
    # print('----- Mutability tests')
    # cr = CharacterRecords[65]
    # try:
    #     cr.Name='Lettre A majuscule'
    #     print('We have a problem, CharacterRecord is mutable.')
    #     print(cr)
    # except AttributeError:
    #     print('CharacterRecord is immutable.')

    print('----- UCDData tests')
    print(GetCodepointFromName('BOAR'))
    print(AsString(0xCAFE))
    print(GetName(0xCAFE))
    print(GetCategory(0xCAFE))
    print(GetScript(0xCAFE))
    print(IsValidCodepoint(0x_BAD_CAFE))
    print(IsSurrogate(0xCAFE))
    print('|'+GetCodepointHexa(0xE)+'|'+GetName(0xE))
    print('|'+GetCodepointHexa(0x1D)+'|'+GetName(0x1D))
    print('|'+GetCodepointHexa(0x99)+'|'+GetName(0x99))
    print('|'+GetCodepointHexa(0xCAFE)+'|'+GetName(0xCAFE))
    print('|'+GetCodepointHexa(0x1CAFE)+'|'+GetName(0x1CAFE))
    print('|'+GetCodepointHexa(0x10CAFE)+'|'+GetName(0x10CAFE))

File: 10_UniView/CS_FW4.8/UniData.cs
// Class UniData.cs
// Reads and stores Unicode UCD file UnicodeData.txt (currently use version from Unicode 11)
// Provides official name and category for each valid codepoint. Not restricted to BMP.
//
// 2018-08-30   PV
// 2020-09-09   PV      1.2: .Net FW 4.8, UnicodeData.txt as embedded resource, UnicodeVersion.txt, Unicode 13
// 2020-12-14   PV      1.5.3: Name override for some ASCII control characters
// 2020-12-14   PV      1.5.4: NonCharacters added manually to charname_map for specific naming
// 2020-12-30   PV      Getting closer to the equivalent in UniSearch.  Renamed from UnicodeData to UniData.  Added scripts info.

using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Diagnostics;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Text;

//#pragma warning disable IDE0057 // Use range operator: Not supported by C#8
#nullable enable

namespace UniViewNS
{
    public class CharacterRecord
    {
        /// <summary>Unicode Character Codepoint, between 0 and 0x10FFFF (from UnicodeData.txt).</summary>
        public int Codepoint { get; private set; }

        /// <summary>Unicode Character Name, uppercase string such as LATIN CAPITAL LETTER A (from UnicodeData.txt).</summary>
        public string Name { get; private set; }

        /// <summary>Unicode General Category, 2 characters such as Lu (from UnicodeData.txt).</summary>
        public string Category { get; private set; }

        /// <summary>Unicode script (from Scripts.txt)</summary>
        public string Script { get => script ?? "Unknown"; }
        internal string? script = null;

        /// <summary>When true, Character method will return an hex codepoint representation instead of the actual string.</summary>
        public bool IsPrintable { get; private set; }

        public CharacterRecord(int cp, string name, string cat, bool isPrintable) => (Codepoint, Name, Category, IsPrintable) = (cp, name, cat, isPrintable);
    }

    public static class UniData
    {
        private static string _UnicodeVersion = "";

        // Real internal dictionary used to store Unicode data
        private static readonly Dictionary<int, CharacterRecord> char_map = new Dictionary<int, CharacterRecord>(65536);

        // Public read only dictionary to access Unicode data

        /// <summary>Dictionary of all valid character records indexed by Codepoints.</summary>
        public static ReadOnlyDictionary<int, CharacterRecord> CharacterRecords { get; } = new ReadOnlyDictionary<int, CharacterRecord>(char_map);

        /// <summary>Last valid codepoint.</summary>
        public const int MaxCodepoint = 0x10FFFF;

        /// <summary>Static constructor, loads data from resources</summary>
        static UniData()
        {
            Stopwatch TotalStopwatch = Stopwatch.StartNew();
            Stopwatch UnicodeDataStopwatch = ReadUnicodeData();
            Stopwatch ScriptsStopwatch = ReadScripts();
            ReadUnicodeVersion();
            TotalStopwatch.Stop();

            Debug.WriteLine("UniData initialization times:");
            Debug.WriteLine($"UnicodeData: {UnicodeDataStopwatch.Elapsed}");
            Debug.WriteLine($"Scripts:     {ScriptsStopwatch.Elapsed}");
            Debug.WriteLine($"TOTAL:       {TotalStopwatch.Elapsed}");
        }

        /// <summary>Read characters info from UnicodeData.txt: Name, Category, IsPrintable</summary>
        private static Stopwatch ReadUnicodeData()
        {
            Stopwatch UnicodeDataStopwatch = Stopwatch.StartNew();
            using (var sr = new StreamReader(GetResourceStream("UnicodeData.txt"), Encoding.UTF8))
                while (!sr.EndOfStream)
                {
                    string[] fields = (sr.ReadLine() ?? string.Empty).Split(';');
                    int codepoint = int.Parse(fields[0], NumberStyles.HexNumber);
                    string char_name = fields[1];
                    string char_category = fields[2];

                    // Special name overrides
                    if (codepoint == 28)
                        char_name = "CONTROL - FILE SEPARATOR";
                    else if (codepoint == 29)
                        char_name = "CONTROL - GROUP SEPARATOR";
                    else if (codepoint == 30)
                        char_name = "CONTROL - RECORD SEPARATOR";
                    else if (codepoint == 31)
                        char_name = "CONTROL - UNIT SEPARATOR";
                    else if (codepoint < 32 || (codepoint >= 0x7f && codepoint < 0xA0))
                        char_name = "CONTROL - " + (fields[10].Length > 0 ? fields[10] : fields[0].Substring(2));

                    bool is_range = char_name.EndsWith(", First>", StringComparison.OrdinalIgnoreCase);
                    bool is_printable = !(codepoint < 32                                // Control characters 0-31
                                        || (codepoint >= 0x7f && codepoint < 0xA0)        // Control characters 127-160
                                        || (codepoint >= 0xD800 && codepoint <= 0xDFFF)   // Surrogates
                                        || codepoint == 0x2028                          // U+2028  LINE SEPARATOR
                                        || codepoint == 0x2029                          // U+2029  PARAGRAPH SEPARATOR
                                        );
                    if (is_range)   // Add all characters within a specified range
                    {
                        char_name = char_name.Replace(", First>", string.Empty).Replace("<", string.Empty).ToUpperInvariant(); //remove range indicator from name
                        fields = (sr.ReadLine() ?? string.Empty).Split(';');
                        int end_char_code = int.Parse(fields[0], NumberStyles.HexNumber);
                        if (!fields[1].EndsWith(", Last>", StringComparison.OrdinalIgnoreCase))
                            Debugger.Break();
                        // Skip planes 15 and 16 private use
                        if (codepoint != 0xF0000 && codepoint != 0x100000)
                            for (int code = codepoint; code <= end_char_code; code++)
                                char_map.Add(code, new CharacterRecord(code, $"{char_name}-{code:X4}", char_category, is_printable));
                    }
                    else
                    {
                        if (codepoint < 0x20000)
                            char_map.Add(codepoint, new CharacterRecord(codepoint, char_name, char_category, is_printable));
                    }
                }

            // Add missing non-characters
            static void AddNonCharacter(int codepoint) => char_map.Add(codepoint, new CharacterRecord(codepoint, $"Not a character - {codepoint:X4}", "", false));

            // 2 last characters of each plane
            for (int plane = 0; plane <= 16; plane++)
            {
                AddNonCharacter((plane << 16) + 0xFFFE);
                AddNonCharacter((plane << 16) + 0xFFFF);
            }
            // FDD0..FDEF: 16 non-characters in Arabic Presentation Forms-A
            for (int code = 0xFDD0; code <= 0xFDEF; code++)
                AddNonCharacter(code);

            return UnicodeDataStopwatch;
        }

        /// <summary>Read script associated with characters from Scripts.txt.</summary>
        private static Stopwatch ReadScripts()
        {
            Stopwatch ScriptsStopwatch = Stopwatch.StartNew();
            using (var sr = new StreamReader(GetResourceStream("Scripts.txt"), Encoding.UTF8))
                while (!sr.EndOfStream)
                {
                    string line = sr.ReadLine() ?? string.Empty;
                    if (line.Length == 0 || line[0] == '#') continue;
                    int p = line.IndexOf('#');
                    if (p >= 0) line = line.Substring(0, p - 1);
                    string[] fields = line.Split(';');
                    p = fields[0].IndexOf('.');
                    int codepoint = int.Parse(p < 0 ? fields[0] : fields[0].Substring(0, p), NumberStyles.HexNumber);
                    string script = fields[1].Trim();
                    if (p < 0)
                    {
                        if (char_map.ContainsKey(codepoint))
                            char_map[codepoint].script = script;
                    }
                    else
                    {
                        int end_char_code = int.Parse(fields[0].Substring(p + 2), NumberStyles.HexNumber);
                        for (int code = codepoint; code <= end_char_code; code++)
                            if (char_map.ContainsKey(code))
                                char_map[code].script = script;
                    }
                }

            ScriptsStopwatch.Stop();
            return ScriptsStopwatch;
        }

        /// <summary>Read version information from UnicodeVersion.txt.  This file is manually managed.</summary>
        private static void ReadUnicodeVersion()
        {
            using var sr = new StreamReader(GetResourceStream("UnicodeVersion.txt"), Encoding.UTF8);
            _UnicodeVersion = sr.ReadLine() ?? "Unknown version";
        }

        /// <summary>Returns stream from embedded resource name.</summary>
        private static Stream GetResourceStream(string name)
        {
            name = "." + name;
            var assembly = typeof(UniData).GetTypeInfo().Assembly;
            var qualifiedName = assembly.GetManifestResourceNames().FirstOrDefault(n => n.EndsWith(name, StringComparison.OrdinalIgnoreCase));
            if (qualifiedName == null)
                throw new ArgumentException("Can't get resource (#1) " + name);
            var st = assembly.GetManifestResourceStream(qualifiedName);
            if (st == null)
                throw new ArgumentException("Can't get resource (#2) " + name);
            return st;
        }

        internal static string GetUnicodeVersion() => _UnicodeVersion;

        public static int GetCPFromName(string name) => char_map.Values.FirstOrDefault(c => string.Compare(c.Name, name, true) == 0)?.Codepoint ?? -1;

        /// <summary>
        /// Converts a codepoint to an UTF-16 encoded string (.Net string).
        /// Do not name it ToString since it's a static class and it can't override object.ToString(.)
        /// </summary>
        /// <param name="cp">Codepoint to convert.</param>
        /// <returns>A string of one character for cp&lt;0xFFFF, or two surrogate characters for cp&gt;=0x10000.
        /// No check is made for invalid codepoints, returned string is undefined in this case.</returns>
        public static string AsString(int cp) => cp <= 0xD7FF || (cp >= 0xE000 && cp <= 0xFFFF) ? new string((char)cp, 1) : new string((char)(0xD800 + ((cp - 0x10000) >> 10)), 1) + new string((char)(0xDC00 + (cp & 0x3ff)), 1);

        public static string GetName(int cp) => char_map.ContainsKey(cp) ? char_map[cp].Name : $"Unassigned codepoint - {cp:X4}";

        public static string GetCategory(int cp) => char_map.ContainsKey(cp) ? char_map[cp].Category : "??";

        public static string GetScript(int cp) => char_map.ContainsKey(cp) ? char_map[cp].Script : "Unknown";

        // Surrogates D800..DFFF are not valid codepoints despite range being listed in UnicodeData.txt
        // NonCharacters are added manually to charname_map for specific namind despite being invalid codepoints
        public static bool IsValidCodepoint(int cp) => !IsSurrogate(cp) && !IsNonCharacter(cp) && char_map.ContainsKey(cp) && cp <= MaxCodepoint;

        internal static bool IsSurrogate(int cp) => cp >= 0xD800 && cp <= 0xDFFF;

        // Last two characters of each page are "Not a character"
        internal static bool IsNonCharacter(int cp) => (cp >= 0xFDD0 && cp <= 0xFDEF) || (cp & 0xFFFF) == 0xFFFE || (cp & 0xFFFF) == 0xFFFF;
    }
}
